saleor/warehouse/management.py

```diff
@@ -24,13 +26,39 @@ def allocate_stock(
     Iterate by stocks and allocate as many items as needed or available in stock
     for order line, until allocated all required quantity for the order line.
     If there is less quantity in stocks then rise InsufficientStock exception.
+    If type of line is CheckoutLine, function deletes previous allocations if
+    quantity differ from current allocations amount.
     """
     stocks = (
         Stock.objects.select_for_update(of=("self",))
-        .get_variant_stocks_for_country(country_code, order_line.variant)
-        .order_by("pk")
+            .get_variant_stocks_for_country(country_code, line.variant)
+            .order_by("pk")
     )
 
+    if type(line) == CheckoutLine:
+        checkout_allocations = Allocation.objects.filter(stock__in=stocks, checkout_line=line)
+        if checkout_allocations is not None:
+            # checkout_allocations._raw_delete(checkout_allocations.db)
+            checkout_allocations.delete()
+
+    if type(line) == OrderLine and checkout.user:
+        if not checkout.expired():
+            checkout_line = checkout.lines.get(variant=line.variant)
+            if checkout_line:
+                checkout_allocations = Allocation.objects.filter(stock__in=stocks, checkout_line=checkout_line)
+                quantity_allocated_for_checkout = 0
+                for allocation in checkout_allocations:
+                    quantity_allocated_for_checkout += allocation.quantity_allocated
+
+                if quantity != quantity_allocated_for_checkout:
+                    checkout_allocations.delete()
```

Comment:
deleting old allocations might be not best option, but it looks like safest option - updateing allocations on multiple stocks sounds like something which will most likely fail at some point